Generics: Passing an array of generic type without type annotation causes C error

### Describe the bug

Passing an array that generic function returns to other generic function, it causes C error

### Reproduction Steps

```v
fn function_that_receives_an_array_of_generic_type[T](array []T) {
	return
}

fn function_that_returns_an_array_of_generic_type[T]() []T {
	return []
}

fn func[T]() {
	res := function_that_returns_an_array_of_generic_type[T]()
	function_that_receives_an_array_of_generic_type(res)
	return
}

fn main() {
	func[string]()
}
```

### Expected Behavior

Successful compilation, or a V error

### Current Behavior

```
======== Output of the C Compiler (cc) ========
/tmp/v_501/code.01JGBDZRRD9V6ZGP0VAJJZJW53.tmp.c:1315:91: error: unknown type name 'Array_Array_string'; did you mean 'Array_string'?
 1315 | VV_LOCAL_SYMBOL void main__function_that_receives_an_array_of_generic_type_T_Array_string(Array_Array_string __v_array);
      |                                                                                           ^~~~~~~~~~~~~~~~~~
      |                                                                                           Array_string
/tmp/v_501/code.01JGBDZRRD9V6ZGP0VAJJZJW53.tmp.c:778:15: note: 'Array_string' declared here
  778 | typedef array Array_string;
      |               ^
/tmp/v_501/code.01JGBDZRRD9V6ZGP0VAJJZJW53.tmp.c:5559:91: error: unknown type name 'Array_Array_string'; did you mean 'Array_string'?
 5559 | VV_LOCAL_SYMBOL void main__function_that_receives_an_array_of_generic_type_T_Array_string(Array_Array_string __v_array) {
      |                                                                                           ^~~~~~~~~~~~~~~~~~
      |                                                                                           Array_string
/tmp/v_501/code.01JGBDZRRD9V6ZGP0VAJJZJW53.tmp.c:778:15: note: 'Array_string' declared here
  778 | typedef array Array_string;
      |               ^
2 errors generated.
===============================================
```

### Possible Solution

_No response_

### Additional Information/Context

It compiles successfully when you add type annotation:

```diff
fn function_that_receives_an_array_of_generic_type[T](array []T) {
	return
}

fn function_that_returns_an_array_of_generic_type[T]() []T {
	return []
}

fn func[T]() {
	res := function_that_returns_an_array_of_generic_type[T]()
-	function_that_receives_an_array_of_generic_type(res)
+	function_that_receives_an_array_of_generic_type[T](res)
	return
}

fn main() {
	func[string]()
}
```

### V version

V 0.4.9 fb5bac1

### Environment details (OS name and version, etc.)

V full version: V 0.4.9 c4aaa2e.fb5bac1
OS: macos, macOS, 15.2, 24C101
Processor: 8 cpus, 64bit, little endian, Apple M3

getwd: /Users/syobon
vexe: /Users/syobon/.v/v
vexe mtime: 2024-12-30 09:17:55

vroot: OK, value: /Users/syobon/.v
VMODULES: OK, value: /Users/syobon/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.39.5 (Apple Git-154)
Git vroot status: 0.4.9-55-gfb5bac11
.git/config present: true

CC version: Apple clang version 16.0.0 (clang-1600.0.26.6)
emcc version: N/A
thirdparty/tcc status: thirdparty-macos-arm64 713692d4

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzcyNjc0Y2YxNDRmNTg1YWU1MTY1MWIiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.NXAxGXdtRT4ZIsEbfDh8Y4xcBdS5_vB12Uh3ZXvuDSE">Huly&reg;: <b>V_0.6-21747</b></a></sub>

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1992,7 +1992,12 @@ fn (mut c Checker) resolve_comptime_args(func &ast.Fn, node_ ast.CallExpr, concr
 							}
 						}
 					} else if call_arg.expr.obj.ct_type_var == .generic_var {
-						mut ctyp := c.type_resolver.get_type(call_arg.expr)
+						mut ctyp := c.unwrap_generic(c.type_resolver.get_type(call_arg.expr))
+						cparam_type_sym := c.table.sym(c.unwrap_generic(ctyp))
+						param_typ_sym := c.table.sym(param_typ)
+						if param_typ_sym.kind == .array && cparam_type_sym.info is ast.Array {
+							ctyp = cparam_type_sym.info.elem_type
+						}
 						if node_.args[i].expr.is_auto_deref_var() {
 							ctyp = ctyp.deref()
 						}

--- a/vlib/v/tests/generics/generic_fn_array_infer_var_test.v
+++ b/vlib/v/tests/generics/generic_fn_array_infer_var_test.v
@@ -1,0 +1,18 @@
+fn function_that_receives_an_array_of_generic_type[T](array []T) {
+	return
+}
+
+fn function_that_returns_an_array_of_generic_type[T]() []T {
+	return []
+}
+
+fn func[T]() {
+	res := function_that_returns_an_array_of_generic_type[T]()
+	function_that_receives_an_array_of_generic_type(res)
+	function_that_receives_an_array_of_generic_type[T](res)
+	return
+}
+
+fn test_main() {
+	func[string]()
+}

--- a/vlib/v/tests/generics/generic_fn_array_infer_var_test.v
+++ b/vlib/v/tests/generics/generic_fn_array_infer_var_test.v
@@ -1,5 +1,5 @@
 fn function_that_receives_an_array_of_generic_type[T](array []T) {
-	return
+	assert array.len == 0
 }
 
 fn function_that_returns_an_array_of_generic_type[T]() []T {
@@ -9,10 +9,12 @@ fn function_that_returns_an_array_of_generic_type[T]() []T {
 fn func[T]() {
 	res := function_that_returns_an_array_of_generic_type[T]()
 	function_that_receives_an_array_of_generic_type(res)
+	assert res.len == 0
 	function_that_receives_an_array_of_generic_type[T](res)
-	return
 }
 
 fn test_main() {
 	func[string]()
+	func[int]()
+	func[f64]()
 }

